src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Prerendering.cs

```diff
@@ -161,8 +161,23 @@ private async Task WaitForResultReady(bool waitForQuiescence, PrerenderedCompone
         }
         else if (_nonStreamingPendingTasks.Count > 0)
         {
-            // Just wait for quiescence of the non-streaming subtrees
-            await Task.WhenAll(_nonStreamingPendingTasks);
+            await WaitForNonStreamingPendingTasks();
+        }
+    }
+
+    private async Task WaitForNonStreamingPendingTasks()
+    {
+        while (_nonStreamingPendingTasks.Count > 0)
+        {
+            // Create a Task that represents the remaining ongoing work for the rendering process
+            var pendingWork = Task.WhenAll(_nonStreamingPendingTasks);
+
+            // Clear all pending work.
+            _nonStreamingPendingTasks.Clear();
```

Comment:
It took me a minute to be convinced by the thread safety of this but I think it is valid, based on:

 * The code executing here holds the renderer's sync context
 * The only code that would add to `_nonStreamingPendingTasks` would itself have to be on the same sync context

If we wanted to make this more explicitly robust, it would be good to add a call to `Dispatcher.AssertAccess();` at the start of `Renderer.cs`'s `AddPendingTask` method. We could even no-op it in the case where we aren't tracking pending tasks, e.g.:

```cs
    // This is inside src\Components\Components\src\RenderTree\Renderer.cs
    protected virtual void AddPendingTask(ComponentState? componentState, Task task)
    {
        // The pendingTasks collection is only used during prerendering to track quiescence,
        // so will be null at other times.
        if (_pendingTasks is {} tasks)
        {
            Dispatcher.AssertAccess();
            tasks.Add(task);
        }
    }
```

Assuming that change does not break anything, would you mind including it in this PR? Then it's much easier to be confident that nothing will be posting extra tasks in a non-threadsafe way.